cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/FileMetaDataReader.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.logging.meta;
+
+import co.cask.cdap.api.Transactional;
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.data.DatasetContext;
+import co.cask.cdap.api.dataset.DatasetManager;
+import co.cask.cdap.api.dataset.table.Row;
+import co.cask.cdap.api.dataset.table.Table;
+import co.cask.cdap.common.io.RootLocationFactory;
+import co.cask.cdap.data.dataset.SystemDatasetInstantiator;
+import co.cask.cdap.data2.datafabric.dataset.DefaultDatasetManager;
+import co.cask.cdap.data2.dataset2.DatasetFramework;
+import co.cask.cdap.data2.dataset2.MultiThreadDatasetCache;
+import co.cask.cdap.data2.transaction.Transactions;
+import co.cask.cdap.data2.transaction.TxCallable;
+import co.cask.cdap.logging.framework.LogPathIdentifier;
+import co.cask.cdap.logging.write.LogLocation;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.security.impersonation.Impersonator;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.primitives.Longs;
+import com.google.inject.Inject;
+import org.apache.tephra.RetryStrategies;
+import org.apache.tephra.TransactionFailureException;
+import org.apache.tephra.TransactionSystemClient;
+
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * class to read log meta data table
+ */
+public class FileMetaDataReader {
+  private static final byte[] ROW_KEY_PREFIX = LoggingStoreTableUtil.FILE_META_ROW_KEY_PREFIX;
+  private static final Comparator<LogLocation> LOG_LOCATION_COMPARATOR = new Comparator<LogLocation>() {
+    @Override
+    public int compare(LogLocation o1, LogLocation o2) {
+      int cmp = Longs.compare(o1.getEventTimeMs(), o2.getEventTimeMs());
+      if (cmp != 0) {
+        return cmp;
+      }
+      // when two log files have same timestamp, we order them by the file creation time
+      return Longs.compare(o1.getFileCreationTimeMs(), o2.getFileCreationTimeMs());
+    }
+  };
+
+  private final DatasetFramework datasetFramework;
+  private final Transactional transactional;
+  private final RootLocationFactory rootLocationFactory;
+  private final Impersonator impersonator;
+
+  @Inject
+  public FileMetaDataReader(final DatasetFramework datasetFramework,
```

Comment:
For `@Inject` constructor, better to make it non-public (just leave it as package private).